Dev#2

Haremun opened the pull request:
I had problem with commits on git hub but now changes should be correct. Only last commit is new, commits before it are already on master branch. I will squash it to one commit and fix problem with branches.
About 1100 new lines are from automated test log file so don't worry :)

+Added first end to end test (you can run it with ./runAutoTest.sh command)

-There is major bug for horizontal checker. 
-Java docs are missing

##REVIEW CHANGES##
A lot of changes are done after installing checkstyle plugin. I worked
on my packages. Removed test from PositionCalculatorTest after changing
a way that method parseInt works. Now instead throwing excpetion returns
-1 which is number that handles board class.
Added a lot of java docs.
Still there is bug with horizontal win checker. Fixing it is my next
goal.

Haremun committed:
Added Travis CI

* Added travis file

* Added test to fail

* Removed failing test, added os in travis

Haremun committed:
User position input, travis fixes

* Fix for travis os

* Added user input
Simple user input and tests for it
Also added dataProvider for FieldTests.
Next step: add fields to the board depedning of user input

* Player input validator - position

* Getting user input and creating fields
User can now inputs symbols on board
User input is validated and calculated. After user input symbols are
printed to the board.
Messages are missing for now.
Test coverage 67%

* Removed windows from travis because there is no java

Haremun committed:
Review fixes

Haremun committed:
Merge branch 'master' into dev

Haremun committed:
Indexing #1

* Added rows and columns indexing

* Simple game flow. Symbols are changing

*Created index creator to index rows and columns
Build painter function is way too long, needs some cuts.
Board can't be bigger than 1000x1000

Haremun committed:
Simple game flow and logger

Symbols are changing.
Indexes for rows and columns.
Added slf4j logger.

Haremun committed:
Win checker - horizontal

Created Judge class with list of checkers.
Added temporary some packages.

Haremun committed:
Merge pull request #9 from Haremun/win

Win checker - horizontal Travis

Haremun committed:
Fixed conflicts with master

Haremun committed:
Review fixes

Haremun committed:
Fix no argument bug

Haremun committed:
travis update

Haremun committed:
Travis with maven 3.6.3

Haremun committed:
Game ends after win, removed while

Haremun committed:
First end to end test

Using bash script to test game. You can run test with ./runAutoTest.sh
command. There are some improvements for game flow class.

Haremun committed:
Merge branch 'game_flow#2' into dev

Haremun committed:
First end to end test

Created bash script for end to end testing. You can run test with
./runAutoTest.sh command. This time test shows bug with horizontal
checking.
There are some fixes for game flow.

LIttleAncientForestKami commented on auto.sh:
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/expect -f
```
Czemu tak?

Haremun replied:
Faktycznie flaga -f była niepotrzebna. Poza tym dzięki tej komendzie mam dostęp do komend expected.

LIttleAncientForestKami replied:
Chodziło mi o wybór `expect`. Nie, że jest zły, jest całkiem w dechę, ale to nie do końca jest Bash. Spójrz:

```
[tammo@lafk-T470 ~]$ man expect                                                                                                                                                                           [INSERT]
Brak podręcznika dla expect
[tammo@lafk-T470 ~]$ expect                                                                                                                                                                               [INSERT]

Command 'expect' not found, but can be installed with:

sudo snap install expect  # version 5.45-7snap0, or
sudo apt  install expect

See 'snap info expect' for additional versions.
```

LIttleAncientForestKami replied:
Mogę go zainstalować, ale sam fakt, że potrzebuję, znaczy że nie jest on domyślnie obecny. Czy zatem np. nie chcesz zaktualizować prerekwizytów w pliku z instrukcjami?

LIttleAncientForestKami commented on auto.sh:
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/expect -f
+ 
+set timeout -1
```
Nieco magicznie, że -1, czemu?

Haremun replied:
Przeoczenie. Zostało usunięte

LIttleAncientForestKami commented on auto.sh:
```diff
@@ -0,0 +1,16 @@
+expect "*Input row and column:\r" 
+send -- "1 0\r"
+
+expect eof
```
Niezły pomysł, natomiast po 3 tygodniach jak tu wrócisz i będziesz to czytał to będziesz się nielicho zastanawiał co właściwie to robi. Jakiś komentarz czy linijka z `echo` nie byłaby głupia.

Haremun replied:
Dodałem dwa komentarze. W kolejnych migawkach test będzie się zmieniał, więc komentarze również

LIttleAncientForestKami commented on log.txt:
```diff
@@ -0,0 +1,1330 @@
+spawn ./run.sh
```
Czy to jest wygenerowany plik? Jeśli tak - po co go wersjonujemy?

Haremun replied:
Dodałem do pliku .gitignore.

LIttleAncientForestKami replied:
Pamiętaj, by także usunąć z wersjonowania. Samo ignorowanie, jeśli plik jest wersjonowany, nie wystarcza.

LIttleAncientForestKami commented on run.sh:
```diff
@@ -0,0 +1 @@
+java -jar target/ox-0.5.jar 30
```
Proponuję `ox-?.?.jar` albo informację w `pom.xml` przy wersji, że jej zmiana wymaga zmiany pliku `run.sh`.

Haremun replied:
Podoba mi się pomysł z ?. Zostało dodane do projektu.

LIttleAncientForestKami commented on runAutoTest.sh:
```diff
@@ -0,0 +1,2 @@
+./auto.sh >> ./log.txt
+echo "Test is stored in file log.txt"
```
Dorzuciłbym linijkę `vim log.txt` albo chociaż `cat`.

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/GameFlow.java:
```diff
@@ -6,11 +6,13 @@
 
-class GameFlow implements Runnable {
+
+class GameFlow implements Runnable, GameListener {
```
Ja widzę wątki!

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/GameFlow.java:
```diff
@@ -6,11 +6,13 @@
     private Board board;
     private List<Symbol> players = new ArrayList<>();
+    private Symbol currentSymbol;
     private boolean win = false;
```
6 pól w klasie. Niemało.

Haremun replied:
Pól ubędzie po dodaniu klasy z ustawieniami rozgrywki. Na chwilę obecną usunąłem jedno niepotrzebne pole.

LIttleAncientForestKami replied:
OK.

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/GameFlow.java:
```diff
@@ -19,32 +21,53 @@
+        currentSymbol = players.get(0);
+
+        makeTurn(positionCalculator);
+
```
Puste linie na końcu metody zawsze proszę wycinać.

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/GameListener.java:
```diff
@@ -0,0 +1,5 @@
+package com.bieganski.ox;
+
+public interface GameListener {
```
Proszę kwitek tu na GH że te Javadoki nie zostaną zapomniane.

Haremun replied:
Javadoki nie zostaną zapomniane. Jeżeli czegoś nie przeoczyłem teraz powinny być w każdym interfejsie oraz w klasach publicznych.

marcinogo replied:
Wygląda na to, że javadocs zostały dodane.

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/PositionCalculator.java:
```diff
@@ -11,7 +11,7 @@
     int calculatePosition(String userInput) {
         if (!positionValidator.checkString(userInput))
-            throw new IllegalArgumentException("There is no such position!");
+            return - 1;
```
Zwykle taka zmiana jest wspominana w dokumentacji. Było nie było, -1 jako szczególna wartość zasługuje na słowo. Zwłaszcza jak poprzednio leciał wyjątek z komentarzem.

Haremun replied:
Dodałem javadok z opisem co to znaczy.

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/PositionCalculator.java:
```diff
@@ -11,7 +11,7 @@
 //TODO maybe parser?
     int calculatePosition(String userInput) {
```
`Integer.parseInt` to metoda na przerabianie napisu w numer. Jak robimy coś podobnego, to trzymajmy konwencję nazewniczą. Przychylam się do notki TODO.

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/painters/IndexCreator.java:
```diff
@@ -3,7 +3,7 @@
     private int boardSide;
 
-    public IndexCreator(int boardSide) {
+    IndexCreator(int boardSide) {
```
:+1:

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/win_checkers/Judge.java:
```diff
@@ -15,15 +16,22 @@
-        if (checkWin(fieldsWithValue, addedField, boardSize))
+        if (checkWin(fieldsWithValue, addedField, boardSize)) {
             App.LOG.info("Winner is found!");
+            gameListener.onWin();
```
What about `onDraw()`?

Haremun replied:
Added. Actually without any logic.

LIttleAncientForestKami commented on src/main/resources/logback.xml:
```diff
@@ -1,17 +1,7 @@
-        </encoder>
-    </appender>-->
-
+    <statusListener class="ch.qos.logback.core.status.NopStatusListener" />
```
Podczas dema spodziewaj się pytania o coś takiego. :-)

LIttleAncientForestKami commented on src/test/java/com/bieganski/ox/PositionCalculatorTest.java:
```diff
@@ -33,13 +33,13 @@ public void testCalculateProperValue() {
-        positionCalculator.calculatePosition(userInput);
-        //Then throws
-    }
+//    @Test(expectedExceptions = IllegalArgumentException.class,
```
Kwita na to poproszę.

Haremun replied:
Test został usunięty po zmianie logiki w metodzie parsera. Dodałem informację w migawce.

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/GameListener.java:
```diff
@@ -0,0 +1,5 @@
+package com.bieganski.ox;
+
+public interface GameListener {
+    void onWin();
```
Co z `onDraw`?

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/GameFlow.java:
```diff
@@ -19,32 +21,53 @@
     }
 
     @Override
     public void run() {
```
Why threads?

LIttleAncientForestKami replied:
Thanks for the thumbs-up. So... why? :-)

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/GameFlow.java:
```diff
@@ -19,32 +21,53 @@
     public void run() {
+        initializeBoard();
+
         PositionCalculator positionCalculator = new PositionCalculator(new PositionValidator(), boardSize);
```
Ta klasa robi część roboty interfejsu użytkownika. A jest w głównym pakiecie. Hmmm.

Haremun replied:
Dobra uwaga. Zostało zmienione.

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/GameFlow.java:
```diff
@@ -19,32 +21,53 @@
+    }
+
+    private void makeTurn(PositionCalculator positionCalculator) {
+        askPlayerForInputAndPlaceSymbol(positionCalculator, currentSymbol);
```
Nie do końca łapię czemu kalkulowanie pozycji jest potrzebne do pytania użyszkodnika o to gdzie stawia krzyżyk. :/

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/GameFlow.java:
```diff
@@ -19,32 +21,53 @@
+    }
+
+    private void askPlayerForInputAndPlaceSymbol(PositionCalculator positionCalculator, Symbol currentSymbol) {
+        userInterface.println("Input row and column:");
```
Wielojęzyczność nie będzie prosta do wprowadzenia. :-)

Haremun replied:
Będę nad tym pracował w kolejnych migawkach.

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/GameFlow.java:
```diff
@@ -19,32 +21,53 @@
+    private void askPlayerForInputAndPlaceSymbol(PositionCalculator positionCalculator, Symbol currentSymbol) {
+        userInterface.println("Input row and column:");
+        String input = userInterface.askForInput();
+        boolean added = board.addField(new Field(positionCalculator.calculatePosition(input), currentSymbol));
```
Jaki jest sens zmiennej `added`? Czemu w następnej linii nie podać po prostu `board.addField`?

LIttleAncientForestKami reviewed (changes requested):
Dużo pytań do kodu, trochę propozycji poprawek. Przede wszystkim pytania pod kątem "dlaczego", "po co to".

Haremun committed:
Review changes

A lot of changes are done after installing checkstyle plugin. I worked
on my packages. Removed test from PositionCalculatorTest after changing
a way that method parseInt works. Now instead throwing excpetion returns
-1 which is number that handles board class.
Added a lot of java docs.
Still there is bug with horizontal win checker. Fixing it is my next
goal.

LIttleAncientForestKami commented on .gitignore:
```diff
@@ -149,3 +149,4 @@ buildNumber.properties
 *.iml
 stale_outputs_checked
 logs/
+log.txt
```
:+1:

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/BoardListener.java:
```diff
@@ -1,8 +1,21 @@
 import java.util.TreeSet;
 
+/**
+ * Interface for notifying about changes on the board.
```
Zerknij proszę na ostatniego mejla mojego o jawadokach. Ma sznurki jakie warto przeczytać. Tu np. mamy masło maślane, czyli dokumentację X zaczynającą się od X (tu X == interface).

Haremun replied:
Racja, zostało poprawione.

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/BoardListener.java:
```diff
@@ -1,8 +1,21 @@
+  //TODO remove size and fieldsWithValue, think about adding placed symbol
+
+  /**
+   * Notify about new field on the board.
```
Notifies?

Haremun replied:
Poprawione

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/checkers/HorizontalWinChecker.java:
```diff
@@ -0,0 +1,47 @@
+      field = correct.first();
+      nextField = fieldsWithValue.lower(field);
+      counter = counter + appendFieldIfCorrect(field, nextField, correct);
+
```
Puste linie po co?

Haremun replied:
Usunięte.

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/checkers/Judge.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.TreeSet;
+
+/**
+ * Board checker, Judge initialize win checkers
```
Czemu w dwu liniach? I czemu zaczynamy od Board checker? Jak rozumiem, Sędzia i Sprawdzacz Planszy inicjalizują Sprawdzacze Wygranej? Razem? Jak to robią? Jeden woła jedną metodę, drugi drugą? Nie mógłby jeden tego robić? Po co są dwa?

Haremun replied:
Faktycznie nie potrzebny był Board checker. Javadoc poprawiony.

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/checkers/Judge.java:
```diff
@@ -0,0 +1,45 @@
+  private GameListener gameListener;
+
+  /**
+   * Constructor for Judge class.
```
Konstruktor kopiujący nie powinien mieć dokumentacji, chyba, że robi coś nietypowego. Parametry udokumentowane napisaniem swej nazwy raz jeszcze to nie jest "coś nietypowego".

Haremun replied:
Javadoc został usunięty

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/checkers/WinChecker.java:
```diff
@@ -0,0 +1,14 @@
+import java.util.TreeSet;
+
+/**
+ * Interface for win-check classes.
```
Jak w poprzednim przypadku.

Haremun replied:
Poprawione.

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/model/Field.java:
```diff
@@ -0,0 +1,70 @@
+package com.bieganski.ox.model;
+
+/**
+ * Field represents symbol O or X on given by user position.
```
Grammar: on a given.

Haremun replied:
Done.

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/model/Field.java:
```diff
@@ -0,0 +1,70 @@
+  }
+
+  /**
+   * Check if field is in given bounds.
```
Bounds - plural.
int bounds - plural in name, singular in reality.
The second bound is known only if you read the code. 
This is the issue docs here should have addressed.

Haremun replied:
You are right. Done.

LIttleAncientForestKami commented on src/main/java/com/bieganski/ox/model/Field.java:
```diff
@@ -0,0 +1,70 @@
+/**
+ * Field represents symbol O or X on given by user position.
+ */
+public class Field implements Comparable<Field> {
```
I salute the effort of having docs on public methods here, but I'd like the docs to be present on methods if they add something. Like the bounds case (next comment). If a method is clear-cut like most here, docs are redundant.

marcinogo replied:
I think this was resolve by Kamil.

Haremun committed:
Update readme

Updated steps to run program and end to end tests.

Haremun committed:
Review fixes

Fixed some typo and javadocs

marcinogo reviewed (changes requested):
Add inheritance of javadocs in interfaces implementations.

Haremun committed:
Fixes for javadocs

Added inheritDoc to interfaces implementations.

Haremun commented:
Merging pull request after approval by @marcinogo. Please create issue if something was unresolved.
